### ISSUE TITLE
OnInvoke Slow Performance on Local Machine

### ISSUE THREAD

**Issue by XavierGeerinck** [dapr/components-contrib#311]
/area runtime

## What version of Dapr?

> 0.6.0

## Expected Behavior

I **expect ms** performance when `Invoking` a service through the **gRPC** binding

## Actual Behavior

Currently I see ~1s delay between call starting and ending

## Steps to Reproduce the Problem

I'm including a Python example here, but it's not the language itself causing the issue. When logging, I see the following output:

```
== APP == [Client][DaprInvoke][debug] Creating Envelope 2020-04-20 14:10:47.693447
== APP == [Client][DaprInvoke][debug] Creating Envelope 2 2020-04-20 **14:10:47.693627**
== APP == [Client][DaprInvoke][debug] Creating Envelope 3 2020-04-20 **14:10:48.698124**
```

When analysing this in more detail, we can see that between Envelope 2 and 3 there is a delay of around 1000ms happening. This is happening while executing the `client.InvokeService(envelope)` function. 

After logging on the **server** to see for execution and response, the following was generated:

```
== APP == OnInvoke: debug -- 2020-04-20 14:10:48.696802
== APP ==         Finished @ 2020-04-20 14:10:48.697117
```

Which allows me to conclude the following:
1. The server response in `ms` magnitude
2. More importantly, the server start processing the message after around 1s (exactly it start processing at `14:10:48.696802` while the message was generated at `14:10:47.693627`)

The above makes me suspect something happening in the Dapr engine which - I assume - is locating the service before it is able to invoke a method on it. **Is someone able to assist with this?**

To reproduce, the following example can be utilized: https://github.com/dapr/python-sdk/tree/master/examples/invoke-simple and modifying it to include logging through the code included below.

```python
from datetime import datetime
import logging
logging.basicConfig(level=logging.INFO, format='%(message)s')
logger = logging.getLogger('MyLogger')


# Start a gRPC client
port = os.getenv('DAPR_GRPC_PORT')
channel = grpc.insecure_channel(f"localhost:{port}")
client = dapr_services.DaprStub(channel)
print(f"Started gRPC client on DAPR_GRPC_PORT: {port}")

data = Any(value='ACTION=1'.encode('utf-8'))
logger.info(f"[Client][DaprInvoke][debug] Creating Envelope {datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')}")
envelope = dapr_messages.InvokeServiceEnvelope(id="openai-server", method="debug", data=data)
logger.info(f"[Client][DaprInvoke][debug] Creating Envelope 2 {datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')}")
response = client.InvokeService(envelope)
logger.info(f"[Client][DaprInvoke][debug] Creating Envelope 3 {datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')}")
print(response.data.value)

channel.close()
```

**Comment by yaron2** [comment]
I moved the issue to components-contrib where the mDNS resolver resides.

**Comment by benc-uk** [comment]
Just a FYI - I also see the same behavior when making HTTP invoke calls, not using gRPC

**Comment by thebillkidy** [comment]
Any update on this? Currently still there when using normal `dapr run` in runtime version `0.7.1`

**Comment by youngbupark** [comment]
> Any update on this? Currently still there when using normal `dapr run` in runtime version `0.7.1`

I found the bug in mdns service discovery logic. Please look at the fix #362

**Comment by XavierGeerinck** [comment]
> > Any update on this? Currently still there when using normal `dapr run` in runtime version `0.7.1`
> 
> I found the bug in mdns service discovery logic. Please look at the fix #362

That looks awesome! Is there any way I can test this myself easily?

**Comment by youngbupark** [comment]
> > > Any update on this? Currently still there when using normal `dapr run` in runtime version `0.7.1`
> > 
> > 
> > I found the bug in mdns service discovery logic. Please look at the fix #362
> 
> That looks awesome! Is there any way I can test this myself easily?

For selfhost mode, you can download binaries from 
 https://github.com/dapr/dapr/actions/runs/137600612
Then replace `daprd` in /usr/local/bin/ with the new downloaded binary.

For Kubernetes mode, use helm.
```
helm install dapr dapr/dapr -n dapr-system --set-string global.tag=edge
```
